Fix DNS record updates for fog-core 1.25.0 and later

a13m opened the pull request:
When PUT is called for a Dynect CNAME record without specifying
a record_id, the response contains an array of records.  The
original code for supporting CNAME updates assumed that a
single record would be returned.  In fog-core prior to 1.25.0,
merge_attributes was quietly mishandling this data without
raising an exception.  Starting with version 1.25.0, the use
of 'each_pair' to iterate the hash caused an exception to be
raised when an array was passed instead.

This change fixes PUT calls by first retrieving the existing
record and performing the update on the correct record_id.

plribeiro3000 commented:
Hey @a13m. could you add some tests around this?

a13m commented:
@plribeiro3000, I will try to come up with a few.  It'll probably be a few days before I can get back to it.

plribeiro3000 commented:
Sure. Take your time!

a13m commented:
@plribeiro3000 three months later, I still haven't had much time to learn the test framework.  I tried a bit today, but I'm not quite sure how it works.  It seems like what you want me to write is tests under `tests/dns/models/record_tests.rb`, but as far as I can see, both in my local tests and your Travis tests (https://travis-ci.org/fog/fog-dynect/jobs/77057343 for example), the model tests always just show "pending", and I haven't figured out what that means.  The only tests that run/succeed are in `tests/dynect/requests/dns/dns_tests.rb`, which isn't the right place for my tests.

Could you please help get me going in the right direction, so that I can stop making everyone on my team hand-patch this code?   (I'm pretty certain we must be the only people using this code path, given that @blentz contributed it and it's currently completely broken)

plribeiro3000 commented:
@a13m Sorry for the delay to answer here.

I believe the test says pending because it is running live tests but has not keys to do so.

I believe this project might be missing some files. The original [`helper`](https://github.com/fog/fog/blob/master/tests/helper.rb) file at fog has a lot more code.

Can you give it a try?

afrank commented:
I ran into this very same issue and was able to get around it by applying this patch. +1

afrank commented:
@plribeiro3000 would it be possible to get this PR merged? I'm blocked by this issue.

plribeiro3000 commented on lib/fog/dynect/models/dns/record.rb:
```diff
@@ -59,6 +51,25 @@ def zone
+              :type     => tokens[2][0...-6] # everything before 'Record'
+            }
+          end
+          record = records.find(ifnone={}) {|record| record[:type] == type}
```
What does `records.find(ifnone={})` mean?

a13m replied:
From http://ruby-doc.org/core-1.9.3/Enumerable.html#method-i-find

> If no object matches, calls ifnone and returns its result when it is specified...

It's been a while, but I believe this was an attempt to prevent passing a `nil` to `merge_attributes`.  However, the value passed to ifnone would need to be callable for that to work; something like: `records.find(ifnone=Proc.new { {} })`

That code may be left over from an intermediate attempt at fixing this, before I understood the real problem.  I can test removing that and make sure it all still works.

plribeiro3000 replied:
I'm not sure you need to say `(ifnone={})`. Perhaps just `({})` would be enough?

plribeiro3000 replied:
Can you confirm this works without the ifnone part so i can merge it and move forward?

plribeiro3000 commented on lib/fog/dynect/models/dns/record.rb:
```diff
@@ -25,7 +25,9 @@ def save(replace=false)
 
-          if replace
+          refresh_id if replace && identity.nil?
+          if replace && !identity.nil?
```
I guess you can this more readable by doing a nested if like:

``` ruby
if replace 
  if identity.nil?
    refresh_id
  else
    .
    .
    .
  end
end
```

a13m replied:
The else block at line 32 is executed under different conditions than refresh_id, so it would have to be something like:

```
if replace
  if identity.nil?
    refresh_id
    data = service.post_record(type, zone.identity, name, rdata, options).body['data']
  else
    options['record_id'] = identity
    data = service.put_record(type, zone.identity, name, rdata, options).body['data']
  end
else
  data = service.post_record(type, zone.identity, name, rdata, options).body['data']
end
```

If you think that is more clear, then I can change it, but I really did not want to duplicate the post_record call.

plribeiro3000 replied:
I think you are right then. The whole problem is having a boolen param that change the behaviour of the method. It makes it hard to follow and understand. 

Maybe in a future refactoring and major bump we can change that to be two different methods that completely express its behaviour e.g (`save`, `overwrite`).

a13m replied:
Actually, my original comment was not entirely correct.  The real reason the logic was separate is because after `refresh_id` is called, `identity` may no longer be nil.  That was the main point of the patch.  :-)

plribeiro3000 commented:
@afrank I did add some comments here that are not blockers but would help me understand what behind this code since i don't have much knowledge about this codebase beside some parts of the code and there is no tests for it.

Commit:
Fix DNS record updates for fog-core 1.25.0 and later

When PUT is called for a Dynect CNAME record without specifying
a record_id, the response contains an array of records.  The
original code for supporting CNAME updates assumed that a
single record would be returned.  In fog-core prior to 1.25.0,
merge_attributes was quietly mishandling this data without
raising an exception.  Starting with version 1.25.0, the use
of 'each_pair' to iterate the hash caused an exception to be
raised when an array was passed instead.

This change fixes PUT calls by first retrieving the existing
record and performing the update on the correct record_id.

a13m commented on lib/fog/dynect/models/dns/record.rb:
```diff
@@ -59,6 +51,25 @@ def zone
+              :type     => tokens[2][0...-6] # everything before 'Record'
+            }
+          end
+          record = records.find() {|record| record[:type] == type}
```
removed the ifnone.  my use case works without it.

plribeiro3000 commented:
@a13m Done. But i currently don't have access to release `fog-dynect` so i will try to summom @geemus here.

geemus commented:
released in v0.0.3
